[Yeo Yu Jie] iP

yeoyujie opened the pull request:
# SoCCat
> “All great and precious things are lonely.” —John Steinbeck

SoCCat is your friendly neighbourhood chatbot. It could
- text-based
- easy to use
- pretty cute

You simply need to:
1. download it from here
2. double-click it
3. add your tasks.
4. let it do its magic :smirk:

And it is **FREE!**

Features:
- [x] Managing tasks
- [ ] Managing deadlines
- [ ] Managing events

If you Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

yeoyujie committed:
Level-1

yeoyujie committed:
Level-2

yeoyujie committed:
Level 3

yeoyujie committed:
Level-4

yeoyujie committed:
actual vs expected match

yeoyujie committed:
created new Exception classes

yeoyujie committed:
changed from array to arraylist

yeoyujie committed:
finished adding delete function

yeoyujie committed:
test

yeoyujie committed:
create new directory if does not exist

yeoyujie committed:
added loadToDisk

yeoyujie committed:
handled errors from loadFromDisk

yeoyujie committed:
Merge pull request #1 from yeoyujie/branch-Level-7

Branch level 7

yeoyujie committed:
Tidied up load and save

yeoyujie committed:
Merge pull request #2 from yeoyujie/branch-Level-7

Tidied up load and save

yeoyujie committed:
Finish formatting datetime for all tasks

yeoyujie committed:
Merge pull request #3 from yeoyujie/branch-Level-8

Finish formatting datetime for all tasks

yeoyujie committed:
Add gradle files

yeoyujie committed:
Format code to more OOP

yeoyujie committed:
Debug LocalDateTime error due to early parsing

yeoyujie committed:
sort into packages

yeoyujie committed:
Fix null task added bug

yeoyujie committed:
Write unit test cases for Parser and Ui

yeoyujie committed:
Change gradle version

yeoyujie committed:
Add Javadoc to non-private classes/methods

yeoyujie committed:
Edit code to follow  coding standard

yeoyujie committed:
Add find function

PangKuangWei commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,166 @@
+                    deleteTask(words);
+                } else {
+                    throw new DukeInvalidException();
+                }
```
Could consider replacing this with a `switch`.

PangKuangWei commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,23 @@
+    private String[] formattedInput;
+    
+    public Parser() {
+    }
```
Could consider changing all the methods to Class methods.

PangKuangWei commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,90 @@
+                    writer.write(((Events) task).getDuration().format(DATE_TIME_FORMATTER));
+                }
+                writer.write("\n");
+            }
```
Consider using FileWriter's `append` method.

PangKuangWei commented on src/main/java/duke/exception/DukeDateTimeException.java:
```diff
@@ -0,0 +1,8 @@
+package duke.exception;
+
+public class DukeDateTimeException extends DukeException {
+    
+    public DukeDateTimeException() {
+        super("Please input a date in the following format: dd/MM/yyyy HH:mm");
+    }
```
I like that there are different type of duke exceptions as it helps with abstraction and removes the need to copy the error message for that specific exception.

PangKuangWei commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,90 @@
+            while ((line = reader.readLine()) != null) { 
+                String[] words = line.split("[|]", 4);
+                String keyword = words[0];
+                String isDone = words[1];
```
Consider changing the type of `isDone` to boolean here rather than below.

PangKuangWei reviewed (commented):
Your code looks good to me! The code is very well structured and therefore easy to read.

yeoyujie committed:
Merge branch-A-CodingStandard into master

yeoyujie committed:
Move scanner init into constructor

yeoyujie committed:
Merge branch-Level-9

yeoyujie committed:
Merge 3 parallel branches together

The following 3 branches are merged and merge conflicts were resolved.
* branch-A-JavaDoc
* branch-A-CodingStandard
* branch-Level-9

yeoyujie committed:
Change from if to switch for keyword command

yeoyujie committed:
Move some components from Duke.java to Ui.java to align with OOP

yeoyujie committed:
added new method in parser to check for number of args

rui-han-crh commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,205 @@
+            String keyword = parser.getKeyword(input);
+
+            try {
+                switch (keyword) {
+                    case "bye":
+                        return;
+                    case "list":
+                        ui.printAllTasks(tasks.getTaskList());
+                        break;
+                    case "todo":
+                        createToDos(words);
+                        break;
+                    case "deadline":
+                        createDeadlines(words);
+                        break;
+                    case "event":
+                        createEvents(words);
+                        break;
+                    case "mark":
+                        mark(words);
+                        break;
+                    case "unmark":
+                        unmark(words);
+                        break;
+                    case "delete":
+                        deleteTask(words);
+                        break;
+                    case "find":
+                        findTask(words);
+                        break;
+                    default:
+                        throw new DukeInvalidException();
+                }
```
I noticed an extra indentation for this switch statement on each case label. Perhaps by unidenting each case label, it would better conform to the coding standards.

rui-han-crh replied:
I noticed a good use of abstraction and the SLAP principle here

rui-han-crh commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,205 @@
+        }
+    }
+
+    private void mark(String[] input) throws DukeException {
+        if (input.length < 2) {
+            throw new DukeEmptyException(input[0]);
+        }
+        try {
+            int taskIndex = parser.getTaskIndex(input);
+            Task markedTask = tasks.getTask(taskIndex);
+            markedTask.markAsDone();
+            ui.printTaskMarked(markedTask);
+            storage.saveToDisk(tasks.getTaskList());
+        } catch (NumberFormatException | IndexOutOfBoundsException ex) {
+            throw new DukeIndexOutOfBoundsException(tasks.getSize());
+        }
+    }
+
+    private void unmark(String[] input) throws DukeException {
+        if (input.length < 2) {
+            throw new DukeEmptyException(input[0]);
+        }
+        try {
+            int taskIndex = parser.getTaskIndex(input);
+            Task unmarkedTask = tasks.getTask(taskIndex);
+            unmarkedTask.unmarkAsNotDone();
+            ui.printTaskUnmarked(unmarkedTask);
+            storage.saveToDisk(tasks.getTaskList());
+        } catch (NumberFormatException | IndexOutOfBoundsException ex) {
+            throw new DukeIndexOutOfBoundsException(tasks.getSize());
+        }
+    }
```
It seems like `Duke::mark` and `Duke::unmark` perform roughly the same procedure with only a differing call to the Ui object. Perhaps we may have a more general function that abstracts out the identical lines, and infer from the input whether to mark or unmark, such that the code can be made more compact?
```suggestion
    ...
    case "mark":
    case "unmark":
        annotate(words);
        break;
    ....
    
    private void annotate(String[] input) throws DukeException {
        if (input.length < 2) {
            throw new DukeEmptyException(input[0]);
        }
        
        try {
            int taskIndex = parser.getTaskIndex(input);
            Task annotatedTask= tasks.getTask(taskIndex);
            if (input[0].equals("mark")) {
                annotatedTask.markAsDone();
                ui.printTaskMarked(annotatedTask);
            } else {
                annotatedTask.unmarkAsNotDone();
                ui.printTaskUnmarked(annotatedTask);
            }
            storage.saveToDisk(tasks.getTaskList());
        } catch (NumberFormatException | IndexOutOfBoundsException ex) {
            throw new DukeIndexOutOfBoundsException(tasks.getSize());
        }
    }
```

rui-han-crh commented on src/test/java/duke/UiTest.java:
```diff
@@ -0,0 +1,28 @@
+    @Test
+    void printTask_emptyList() {
+        Ui ui = new Ui();
+        System.setOut(new PrintStream(outputStream));
```
I like how you were able to redirect your output stream here for testing purposes!

rui-han-crh reviewed (commented):
Looks good ^_^ Function names were explicit and denote clearly their responsibilities and well abstracted functions made it easy to read.

yeoyujie committed:
Add config file for checkstyle

Config folder is created and 2 new files added. \
Modifications were done to most files to follow with the java coding conventions.

yeoyujie committed:
Fix some code style violations

yeoyujie committed:
Add javafx

yeoyujie committed:
Set up Javafx and interact with user

yeoyujie committed:
Add FXML into JavaFX

yeoyujie committed:
Change ui return methods from void to String to
work with the GUI.

yeoyujie committed:
Add line breaks for some methods in ui

yeoyujie committed:
Rename files and remove unused imports

yeoyujie committed:
Change to data.txt

yeoyujie committed:
Add assertions to Duke class

Assertions are added to verify that taskIndex is within range

For example, if the task list contains 4 tasks, the taskIndex inputted
by the user must be between 1 and 4 inclusive.

The following methods have assertions added in Duke class
* mark
* unmark
* deleteTask

yeoyujie committed:
Edit Storage methods to follow code standard

yeoyujie committed:
Merge pull request #5 from yeoyujie/branch-A-Assertions

Branch a assertions

yeoyujie committed:
Merge pull request #6 from yeoyujie/branch-A-CodeQuality

Edit Storage methods to follow code standard

yeoyujie committed:
Fix GUI not showing welcome message

The GUI was not showing welcome message from SoCCat as the component
was not included in the dialogContainer.

User prompt and list of available commands are now shown together with
the welcome message so that it is more user-friendly and users would
know what commands to type in.

yeoyujie committed:
Create undo for mark and unmark

yeoyujie committed:
Add docs for find in TaskList

yeoyujie committed:
Fix the bug where tasks is not saved into file properly

There was missing brackets around the lambda expression. When the task
is marked, the task description is left out. Properly inserted brackets
to fix the issue.

yeoyujie committed:
Edit JavaFX for better GUI

yeoyujie committed:
Add undo method for all tasks

Previously, undoing delete added the deleted task into the end of the
list. After this change, the deleted task is restored to its previous
state.

yeoyujie committed:
Update README.md

yeoyujie committed:
Update README.md

yeoyujie committed:
Update complete README.md

yeoyujie committed:
Add Ui.png

yeoyujie committed:
Merge branch 'master' of github.com:yeoyujie/ip

yeoyujie committed:
Add application close when bye is inputted

yeoyujie committed:
Change build.gradle

yeoyujie committed:
Edit javadoc for public and protected methods

yeoyujie committed:
Fix indentations and checktyle errors
